[3.x] Create THIRD-PARTY-LICENSES.txt file using create-attribution-doc.sh

himani2411 opened the pull request:
### Description of changes
* Add a Bash script to create THIRD_PARTY_LICENSES.txt 
* Added Python License depending on the version where you run this script

### Tests
* Local tests to create a THIRD_PARTY_LICENSES.txt 


### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Create THIRD-PARTY-LICENSES.txt file using create-attribution-doc.sh

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+
+set -ex
+
+pyenv virtualenv attribution-doc-env
```
should we specify the python version here?  This may not be python 3.

himani2411 replied:
Initially i did keep the python version as 3.9.10. But, when I run this script from another virtual environment it takes the Python version used in that environment. We can keep a comment to run this in the latest python version.  Or I could keep this at the latest python 3 version.

dreambeyondorange replied:
I don't think it makes sense to set this up without specifying the version of python.  I'm not sure I understand `when I run this script from another virtual environment it takes the Python version used in that environment`.  Aren't you creating the virtual environment here?

himani2411 replied:
I have specified the python version to 3.9.10

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+
+
+# Update Pip
+pip3 install --upgrade pip
```
If we are using pyenv, shouldn't this be just `pip`?

himani2411 replied:
If I keep the latest python version I think we wouldnt have to worry about this

dreambeyondorange replied:
I don't think we should be using the pyenv and pip3 at the same time.  Either we set up the virtual env and use the package manager for it, or we require 3.x and use pip3

himani2411 replied:
I have specified the python version to 3.9.10. So should resolve the use of pip3

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+python_license_file=python_license.txt
+aws_cdk_license_file=aws-cdk-license.txt
+# Create a list of aws cdk Sub-packages we want to ignore 
+aws_cdk_ignore_subpackages=$(pip list | grep cdk | awk '{print $1}') 
```
Can we safely assume all the packages use the same license?

himani2411 replied:
Yes, all f the sub-packages use the same version and the same license which is why in order to remove the duplication, I have added the aws-cdk license once, which is what we do in the older version of 3rd party license too

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+# Create a list of aws cdk Sub-packages we want to ignore 
+aws_cdk_ignore_subpackages=$(pip list | grep cdk | awk '{print $1}') 
+# Create a pip License document
+pip-licenses -i $aws_cdk_ignore_subpackages aws-parallelcluster pip-licenses --format=plain-vertical --with-license-file --with-urls --no-license-path --with-authors --output-file=$final_license_file
```
Does this also capture transitive dependencies?  I.e. the dependencies of our dependencies?

himani2411 replied:
yes, it does. For example, https://github.com/raimon49/pip-licenses/issues/41

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+cpy_version=$(python -V |  grep -Eo '([0-9]+)(\.?[0-9]+)' | head -1) 
+
+
+# Add other License in the final file
```
I would suggest clarifying this comment a bit more.  What exactly are we doing?

himani2411 replied:
I did add other comments below these, but will elaborate on them on a high level.

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+
+# Add other License in the final file
+# Get Aws-cdk License file
+echo "aws-cdk; https://raw.githubusercontent.com/aws/aws-cdk/main/LICENSE" >> $final_license_file
```
can you get this from the pyenv instead of fetching online?

himani2411 replied:
I would have to take into account the packages and the versions for finding this. I could find this locally, but isnt it straightforward to get it online. Is there a reason I shouldnt?

dreambeyondorange replied:
how do you guarantee that what you are fetching online is the same version you have locally?

himani2411 replied:
I took the URL from the pip-license command which is https://github.com/aws/aws-cdk. But for concatenating I needed the raw content.

dreambeyondorange commented on util/create-attribution-doc.sh:
```diff
@@ -0,0 +1,47 @@
+# Get Aws-cdk License file
+echo "aws-cdk; https://raw.githubusercontent.com/aws/aws-cdk/main/LICENSE" >> $final_license_file
+curl https://raw.githubusercontent.com/aws/aws-cdk/main/LICENSE >> $final_license_file
+# Download Python License file
```
Is this also available locally?

himani2411 replied:
The Python License file location could differ might change for others, I found getting it online much more straitforward.

dreambeyondorange replied:
similarly, how do you know it matches what you have?

himani2411 replied:
I am taking the Python version we have on pyenv and using that to get the LICENSE file directly

dreambeyondorange reviewed (commented):
Nice work!

Commit:
Add latest Python version for virtual enviornment

Adding latest Python version and aws-cdk version to the THIRD-PARTY-LICENSES.txt

PR: https://github.com/aws/aws-parallelcluster/pull/4802

himani2411 committed:
Merge branch 'develop' into wip/create-attribution-doc
